pghoard/basebackup/delta.py

```diff
@@ -0,0 +1,479 @@
+        self.parallel = parallel
+        self.temp_base_dir = temp_base_dir
+        self.compressed_base = compressed_base
+        self.submitted_hashes_lock = threading.Lock()
+        self.submitted_hashes: Set[str] = set()
+        self.tracked_snapshot_files: SnapshotFiles = self._list_existing_files()
+        self.chunk_uploader = chunk_uploader
+        self.data_file_format = data_file_format
+
+    def _snapshot(self, snapshotter: Snapshotter) -> SnapshotResult:
+        snapshotter.snapshot(reuse_old_snapshotfiles=False)
+        snapshot_result = SnapshotResult()
+        snapshot_result.state = snapshotter.get_snapshot_state()
+        snapshot_result.hashes = [
+            SnapshotHash(hexdigest=ssfile.hexdigest, size=ssfile.file_size)
+            for ssfile in snapshot_result.state.files
+            if ssfile.hexdigest
+        ]
+        snapshot_result.files = len(snapshot_result.state.files)
+        snapshot_result.total_size = sum(ssfile.file_size for ssfile in snapshot_result.state.files)
+        snapshot_result.end = now()
+
+        self.log.debug("snapshot result: %r", snapshot_result.json())
+
+        return snapshot_result
+
+    def _list_existing_files(self) -> SnapshotFiles:
+        """Iterate through all manifest files and fetch information about hash files"""
+        all_snapshot_files: Dict[str, SnapshotFile] = {}
+        for backup in self.get_remote_basebackups_info(self.site):
+            if backup["metadata"].get("format") not in {BaseBackupFormat.delta_v1, BaseBackupFormat.delta_v2}:
+                continue
+
+            meta, _ = download_backup_meta_file(
+                storage=self.storage,
+                basebackup_path=os.path.join(self.site_config["prefix"], "basebackup", backup["name"]),
+                metadata=backup["metadata"],
+                key_lookup=lambda key_id: self.site_config["encryption_keys"][key_id]["private"],
+                extract_meta_func=extract_pghoard_delta_metadata
+            )
+            manifest = meta["manifest"]
+            snapshot_result = manifest["snapshot_result"]
+            backup_state = snapshot_result["state"]
+            files = backup_state["files"]
+            for delta_file in files:
+                snapshot_file = SnapshotFile.parse_obj(delta_file)
+                if snapshot_file.hexdigest:
+                    all_snapshot_files[snapshot_file.hexdigest] = snapshot_file
+
+        return all_snapshot_files
+
+    def _delta_upload_hexdigest(
+        self, *, temp_dir: Path, chunk_path: Path, file_obj: SizeLimitedFile, callback_queue: CallbackQueue,
+        relative_path: Path
+    ) -> Tuple[int, int, str, bool]:
+        """Schedule a separate delta file for the upload, calculates the final hash to use it as a name
+         (as hash might change during the backup)"""
+        skip_upload = False
+        start_time = time.monotonic()
+
+        input_size = file_obj.seek(0, os.SEEK_END)
```

Comment:
Existing issue but I guess file size can change while it's being uploaded and instead of reading the size here we should check how many bytes we actually ended up uploading (and if hash matches then checking what's the actual size of the file that matched)?